src/lib/walletManager.ts

```diff
@@ -0,0 +1,182 @@
+import { JsonRpcProvider, Wallet, getAddress } from "ethers";
+import { OvalDiscovery } from "./";
+import { env } from "./env";
+import { retrieveGckmsKey } from "./gckms";
+import { isMochaTest } from "./helpers";
+import { OvalConfigs, OvalConfigsShared } from "./types";
+
+type WalletConfig = {
+    unlockerKey?: string;
+    gckmsKeyId?: string;
+};
+
+type WalletUsed = {
+    walletPubKey: string;
+    targetBlock: number;
+    count: number;
+}
+
+// WalletManager class to handle wallet operations.
+export class WalletManager {
+    private static instance: WalletManager;
+    private ovalDiscovery: OvalDiscovery;
+    private wallets: Record<string, Wallet> = {};
+    private sharedWallets: Map<string, Wallet> = new Map();
+    private sharedWalletUsage: Map<string, Array<WalletUsed>> = new Map();
```

Comment:
Why did you choose to have an Array to track wallet usage rather than a mapping?

For instance, I could imagine a mapping from block->OvalInstance->WalletUsed might be a decent alternative: simple to drop blocks once they've passed (just delete from the map) and it's easy to tell if there's already been a wallet assigned (no array traversal/find).

Reply:
Thanks, that’s a better data structure from the record cleanup. However, initially, I proposed the data structure with this idea in mind:
- We want to use the same unlocker for a given oval instance, regardless of the target block. This approach is also due to a cleanup running every 5 minutes, preventing old assignments from being reused, focusing only on recent assignments.

The reason for this is that multiple unlocks to the same oval instance in target blocks close to each other are likely part of the same backrun. They belong to the same price auction, meaning nonce collisions are not possible if we reuse the unlocker. Do you agree with this?

The previous data structure seemed to work fine with that assumption. However, I do like removing the array and using only mappings that we still need to traverse, for the reason I explained above.

I’ve changed the data structure in the latest commits.